autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java

```diff
@@ -118,74 +125,105 @@ protected void execute() {
-                existingHostGroupSize, configMaxNodeCount, configMinNodeCount, maxAllowedUpScale, maxAllowedDownScale);
+        LOGGER.info("Various counts: hostFqdnsToInstanceId={}, servicesHealthyHostFqdnsToInstanceId: {}, stoppedHostFqdnsToInstanceId: {}, " +
+                        "configMaxNodeCount={}, configMinNodeCount={}, maxAllowedUpScale={}, maxAllowedDownScale={}",
+                existingHostGroupSize, serviceHealthyHostGroupSize, stoppedHostInstanceIds.size(),
+                configMaxNodeCount, configMinNodeCount, maxAllowedUpScale, maxAllowedDownScale);
 
         Optional<Integer> mandatoryUpScaleCount = Optional.of(configMinNodeCount)
                 .filter(mandatoryUpScale -> mandatoryUpScale < 0).map(upscale -> -1 * upscale);
         Optional<Integer> mandatoryDownScaleCount = Optional.of(configMaxNodeCount)
                 .filter(mandatoryDownscale -> mandatoryDownscale < 0).map(downscale -> -1 * downscale);
 
-        YarnScalingServiceV1Response yarnResponse = yarnMetricsClient
-                .getYarnMetricsForCluster(cluster, stackV4Response, policyHostGroup, pollingUserCrn, mandatoryDownScaleCount);
+        Optional<Integer> mandatoryStopStartUpscaleCount = isStopStartEnabled ? Optional.of(stopStartAdjustment)
+                .filter(mandatoryStopStartUpscale -> mandatoryStopStartUpscale > 0) : Optional.empty();
+        Optional<Integer> mandatoryStopStartDownscaleCount = isStopStartEnabled ? Optional.of(stopStartAdjustment)
+                .filter(mandatoryStopStartDownscale -> mandatoryStopStartDownscale < 0).map(Math::abs) : Optional.empty();
+
+        YarnScalingServiceV1Response yarnResponse =
+                yarnMetricsClient.getYarnMetricsForCluster(cluster, stackV4Response, policyHostGroup, pollingUserCrn, mandatoryDownScaleCount);
 
         if (cluster.getUpdateFailedDetails() != null && pollingUserCrn.equals(cluster.getMachineUserCrn())) {
             // Successful YARN API call
             clusterService.setUpdateFailedDetails(cluster.getId(), null);
         }
 
-        int yarnRecommendedScaleUpCount = yarnResponseUtils.getYarnRecommendedScaleUpCount(yarnResponse, policyHostGroup,
-                maxAllowedUpScale, mandatoryUpScaleCount, loadAlertConfiguration.getMaxScaleUpStepSize());
-        List<String> yarnRecommendedDecommissionHosts = yarnResponseUtils.
-                getYarnRecommendedDecommissionHostsForHostGroup(cluster.getStackCrn(), yarnResponse,
-                        hostFqdnsToInstanceId, maxAllowedDownScale, mandatoryDownScaleCount, loadAlertConfiguration.getMaxScaleDownStepSize());
+        int yarnRecommendedScaleUpCount = yarnResponseUtils.getYarnRecommendedScaleUpCount(yarnResponse, policyHostGroup, maxAllowedUpScale,
+                mandatoryUpScaleCount, loadAlertConfiguration.getMaxScaleUpStepSize(), mandatoryStopStartUpscaleCount);
+        List<String> yarnRecommendedDecommissionHosts =
```

Comment:
This code is starting to get a little too complicated, and difficult to understand IMO.

For instance: There's enough scenarios in what this patch is trying to do (and prior to the patch), where we don't need to call YARN APIs at all. Simple scenarios like maxScaleUpCount being adjusted doesn't need a YARN API call.